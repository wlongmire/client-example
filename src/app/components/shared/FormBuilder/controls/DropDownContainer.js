import React from 'react'
import classNames from 'classnames'
import {
  FormGroup,
  DropdownButton,
  MenuItem,
  ControlLabel,
  Tooltip,
  OverlayTrigger,
  FormControl
} from 'react-bootstrap'


class DropDownContainer extends React.Component {
  constructor(props) {
    super(props)

    const name = this.props.data.name 
    this.state = {
      disabled: (this.props.initialParams[name] && this.props.initialParams[name].disabled)?this.props.initialParams[name].disabled:false,
      validationMessage:""
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleChangeEvent  = this.handleChangeEvent.bind(this)
    this.getValidationState = this.getValidationState.bind(this)
  }

  getValidationState() {
    // this bit of code acts as a circuit breaker
    // We want get validation to trigger a validation function and when results are back we need to call it again
    // But this breaks the loops
    if (this.state.isValid) {
      let isValid = this.state.isValid
      this.state.isValid = null
      return isValid
    }

    //only trigger validation if the value changes
    if (this.state.value !== '' &&
        this.props.data.attributes &&
        this.props.data.attributes.validationFunc &&
        this.props.validation[this.props.data.attributes.validationFunc]) {
          
        this.props.validation[this.props.data.attributes.validationFunc](this.props.data.name, this.state.value).then((result)=> {
          this.setState({
            isValid: (result) ? 'success' : 'error',
            validationMessage:(result.status) ? "" : result.message
          })
        })
    }

    if (this.state.isValid !== null) {
      return this.state.isValid
    }

    if (this.props.data.attributes && this.props.data.attributes.validationRegEx) {
      let regex = new RegExp(unescape(this.props.data.attributes.validationRegEx))
      return (regex.test(this.state.value)) ? 'success' : 'error'
    }

  }

  componentWillMount() {
    if (this.props.initialValues[this.props.data.name]) {
      this.handleChange(this.props.initialValues[this.props.data.name])
    } else {
      this.setState({
        value: this.props.data.attributes.options[0].value
      })
    }
  }

  handleChange(value) {
    let option = this.props.data.attributes.options.filter((option) => {
      return option.value == value
    })[0]

    if(option.supplementalquestionIds && option.supplementalquestionIds.length > 0) {
      this.props.handleSupplementTrigger(option.supplementalquestionIds)
    } else {
      this.props.handleSupplementTrigger([])
    }

    // Change dropdown title on select
    this.setState({
      value: option.value
    })

    this.props.handleFormChange()
  }

  handleChangeEvent(e) {
    this.handleChange(e.currentTarget.value)
  }

  render() {
    const tooltip = <Tooltip id={`tooltip_${this.props.data.questionId}`}> {this.props.data.tooltiptext}</Tooltip>

    this.options = this.props.data.attributes.options.map((data) => {
      return <option value={data.value} key={data.optionId}>{data.text}</option>
    })
        
    return(
       <FormGroup validationState={this.getValidationState()} controlId={this.props.data.name}>
         { this.props.data.text && <ControlLabel>{this.props.data.text}</ControlLabel> }
         
         <OverlayTrigger placement='top' overlay={tooltip} trigger={(this.props.data.tooltiptext) ? ['hover', 'focus'] : null}>
           <div className="select">
            <FormControl 
              value={this.state.value}
              className={classNames({'filled':this.state.value}, {disabled:this.state.disabled})}
              disabled={this.state.disabled}
              onChange={this.handleChangeEvent} 
              componentClass="select">
              {this.options}
            </FormControl>
          </div>
        </OverlayTrigger>
       </FormGroup>
    )
  }
}

export default DropDownContainer