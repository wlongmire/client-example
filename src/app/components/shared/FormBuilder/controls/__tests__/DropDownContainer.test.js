/* eslint-env node, jest */

import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { DropDownContainer } from '../DropDownContainer'

const dataProps = {
  attributes: {
    options: [
      { text: 'State', value: '', optionId: '1'},
      { text: 'New York', value: 'New York', optionId: '2'},
      { text: 'New Jersey', value: 'New Jersey', optionId: '3'},
      { text: 'California', value: 'California', optionId: '4'},
    ]
  },
  inputType: 'dropdown-single',
  name: 'projectState',
  placeholder: 'State',
  required: true,
  tooltiptext: 'Please provide as descriptive of a street address as possible.'
}

describe('>>> DropDownContainer Component --- Snapshot', () => {
  it('+++ provide a dropdown field', () => {
    const renderedValue = shallow(
      <DropDownContainer
        user={{ broker: 'test-2222-7fd-b3ff-4fd3-9fc2-e752b9f5b002'}}
        initialValues={{}}
        initialParams={{}}
        data={dataProps}
      />)
    expect(toJson(renderedValue)).toMatchSnapshot()
  })
})

const initalParamsProps = {
  primaryInsuredAddress: { disabled: true },
  primaryInsuredCity: { disabled: true },
  primaryInsuredName: { disabled: true },
  primaryInsuredState: { disabled: true },
  primaryInsuredZipcode: { disabled: true },
  projectAddress: { disabled: true },
  projectCity: { disabled: true },
  projectState: { disabled: true }
}

const initalValuesProps = {
  clearanceStatus: 'pass',
  primaryInsuredAddress: '234234 23423489',
  primaryInsuredCity: '234234 23423489',
  primaryInsuredName: '234234 23423489',
  primaryInsuredState: 'District of Columbia',
  primaryInsuredZipcode: '234234 23423489',
  projectAddress: '234234 23423489',
  projectCity: '234234 23423489',
  projectState: 'California',
  projectZipcode: '234234 23423489',
  status: 'SUBMISSION',
  type: 'oi'
}

describe('>>> DropDownContainer Component (with initial params) --- Snapshot', () => {
  it('+++ provide a dropdown field', () => {
    const renderedValue = shallow(
      <DropDownContainer
        user={{ broker: 'test-2222-7fd-b3ff-4fd3-9fc2-e752b9f5b002' }}
        initialValues={initalValuesProps}
        initialParams={initalParamsProps}
        data={dataProps}
        handleSupplementTrigger={jest.fn}
        handleFormChange={jest.fn}
      />)
    expect(toJson(renderedValue)).toMatchSnapshot()
  })
})