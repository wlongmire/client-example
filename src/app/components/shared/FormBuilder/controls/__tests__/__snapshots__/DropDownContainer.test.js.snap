// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> DropDownContainer Component (with initial params) --- Snapshot +++ provide a dropdown field 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="projectState"
>
  <OverlayTrigger
    defaultOverlayShown={false}
    overlay={
      <Tooltip
        bsClass="tooltip"
        id="tooltip_undefined"
        placement="right"
      >
         
        Please provide as descriptive of a street address as possible.
      </Tooltip>
    }
    placement="top"
    trigger={
      Array [
        "hover",
        "focus",
      ]
    }
  >
    <div
      className="select"
    >
      <FormControl
        bsClass="form-control"
        className="filled disabled"
        componentClass="select"
        disabled={true}
        onChange={[Function]}
        value="California"
      >
        <option
          value=""
        >
          State
        </option>
        <option
          value="New York"
        >
          New York
        </option>
        <option
          value="New Jersey"
        >
          New Jersey
        </option>
        <option
          value="California"
        >
          California
        </option>
      </FormControl>
    </div>
  </OverlayTrigger>
  <AdditionalInfoComponent />
</FormGroup>
`;

exports[`>>> DropDownContainer Component --- Snapshot +++ provide a dropdown field 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="projectState"
>
  <OverlayTrigger
    defaultOverlayShown={false}
    overlay={
      <Tooltip
        bsClass="tooltip"
        id="tooltip_undefined"
        placement="right"
      >
         
        Please provide as descriptive of a street address as possible.
      </Tooltip>
    }
    placement="top"
    trigger={
      Array [
        "hover",
        "focus",
      ]
    }
  >
    <div
      className="select"
    >
      <FormControl
        bsClass="form-control"
        className=""
        componentClass="select"
        disabled={false}
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          State
        </option>
        <option
          value="New York"
        >
          New York
        </option>
        <option
          value="New Jersey"
        >
          New Jersey
        </option>
        <option
          value="California"
        >
          California
        </option>
      </FormControl>
    </div>
  </OverlayTrigger>
  <AdditionalInfoComponent />
</FormGroup>
`;
